### DISCUSSION TITLE
Is there a benchmark of Phlex to Haml, Slim & Erb?

### DISCUSSION THREAD

**Discussion by konung** [yippee-fun/phlex#443] [General]
Hi

Just discovered Phlex through Ruby Weekly :D 

It looks very interesting and reminds me a lot of (LuckyFramework)[https://luckyframework.org/guides/frontend/rendering-html] in Crystal. 👍  and papercraft gem. 

I was just wondering if there is a speed/performance comparison against Slim, Haml, Erb? I personally use Hamlit in my Ruby projects, which is an optimized version of Haml, a) I like Haml's syntax over ERB, b) Performance used to be significantly better over other engines, but now they are getting closer (latest versions of Hamlit, Haml & ERB are in the same ballpark, based Hamlit's benchmark (https://github.com/k0kubun/hamlit/tree/master/benchmark)  

I didn't find anything through Google search or issues/repo here. Very possible, I'm just being blind 😄 

Any pointers to any such comparisons, or benchmarks, would be appreciated. 

Thanks

**Comment by joeldrapper** [comment]
Hey, unfortunately there is nothing recent. I would like to put together a repository with some realistic comparisons but I haven't had the time. Phlex is faster than ActionView mainly because it uses a faster buffer (a raw string) and it doesn't need to do any inflection to look up partials, since they’re constants.

Once the compiler is finished (based on my initial benchmarks), Phlex performance should be about the same as compiled HAML but that's still a while off.

**Comment by konung** [comment]
So after making this pull request  https://github.com/joeldrapper/phlex/pull/481
and thinking more about Real-Life benchmarks https://github.com/joeldrapper/phlex/pull/481#issuecomment-1424659528

I wanted to see how Phlex would perform in a super basic Hanami app (I was just starting to explore newly minted Hanami 2)

Honestly, I'm a bit blown away.

All I'm doing in my Hanami Action (think Rails Controller method/action)

- Generating Timestamp
- Generating random RGB value for `bg_color`
- Taking a `request` object
  - merging it with a timestamp
  - merging  RGB value
- passing `req` && `bg_color` to a template (I tried Haml template, Haml Inline, Phlex & JSON)
- Sending response back

In my template I'm iterating over the request hash object and simply building an HTML table, with randomly generated color for background 
Obviously, JSON has no use for bg_color, but it's there for comparison. 

JSON response is about 1.15x slower than Phlex response 🚀  
Phlex's response is 7.74x (!!!) faster than Haml ! 🚀  


🤯 🤯 🤯 🚀 😄

**Comment by konung** [comment]
Set it up in a repo, and reran results, similar to what I mentioned above, although a bit slower, as my computer went into "take off mode" as fans, started running like crazy. 

Check out the results here: https://github.com/konung/hanami-phlex-test-idea

**Comment by joeldrapper** [comment]
Hey 👋 Thanks for doing all these benchmarks!

Phlex is definitely slower than pure compiled templating languages like Slim, ERB and HAML. But view frameworks like ActionView and ViewComponent add a lot of overhead beyond just rendering the template and that's where Phlex pulls ahead, as you've seen from the Hanami benchmarks.

I’m really happy with the performance so far but I’m also working on a compiler for Phlex that will make the template rendering significantly faster (like 10 - 100x) faster. https://github.com/joeldrapper/phlex-compiler

I’ve had to take a break from this to focus on other things but I’m really excited to get back to it.

**Comment by joeldrapper** [comment]
Here are the results of running this on Ruby 3.2.1 with YJIT on my M1 MacBook Pro with 16GBs of RAM. Phlex got over 10k IPS.

```
hey -c 10 -z 10s http://localhost:2300/phlex

Summary:
  Total:	10.0015 secs
  Slowest:	0.2795 secs
  Fastest:	0.0003 secs
  Average:	0.0010 secs
  Requests/sec:	10309.8967

  Total data:	171439235 bytes
  Size/request:	1662 bytes

Response time histogram:
  0.000 [1]	|
  0.028 [102881]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.056 [120]	|
  0.084 [58]	|
  0.112 [21]	|
  0.140 [23]	|
  0.168 [0]	|
  0.196 [2]	|
  0.224 [0]	|
  0.252 [4]	|
  0.279 [4]	|


Latency distribution:
  10% in 0.0003 secs
  25% in 0.0004 secs
  50% in 0.0005 secs
  75% in 0.0010 secs
  90% in 0.0014 secs
  95% in 0.0018 secs
  99% in 0.0058 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0003 secs, 0.2795 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0034 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0041 secs
  resp wait:	0.0009 secs, 0.0003 secs, 0.2743 secs
  resp read:	0.0000 secs, 0.0000 secs, 0.0048 secs

Status code distribution:
  [200]	103114 responses
```

```
hey -c 10 -z 10s http://localhost:2300/haml_template

Summary:
  Total:	10.0026 secs
  Slowest:	0.2619 secs
  Fastest:	0.0019 secs
  Average:	0.0051 secs
  Requests/sec:	1966.2921

  Total data:	41451668 bytes
  Size/request:	2107 bytes

Response time histogram:
  0.002 [1]	|
  0.028 [19420]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.054 [134]	|
  0.080 [57]	|
  0.106 [24]	|
  0.132 [24]	|
  0.158 [3]	|
  0.184 [3]	|
  0.210 [1]	|
  0.236 [0]	|
  0.262 [1]	|


Latency distribution:
  10% in 0.0026 secs
  25% in 0.0030 secs
  50% in 0.0037 secs
  75% in 0.0049 secs
  90% in 0.0066 secs
  95% in 0.0086 secs
  99% in 0.0355 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0019 secs, 0.2619 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0026 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0024 secs
  resp wait:	0.0050 secs, 0.0019 secs, 0.2563 secs
  resp read:	0.0001 secs, 0.0000 secs, 0.0385 secs

Status code distribution:
  [200]	19668 responses
```

```
hey -c 10 -z 10s http://localhost:2300/haml_inline

Summary:
  Total:	10.0054 secs
  Slowest:	0.3048 secs
  Fastest:	0.0019 secs
  Average:	0.0051 secs
  Requests/sec:	1958.7395

  Total data:	41185878 bytes
  Size/request:	2101 bytes

Response time histogram:
  0.002 [1]	|
  0.032 [19400]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.062 [113]	|
  0.093 [40]	|
  0.123 [20]	|
  0.153 [22]	|
  0.184 [0]	|
  0.214 [0]	|
  0.244 [0]	|
  0.275 [1]	|
  0.305 [1]	|


Latency distribution:
  10% in 0.0027 secs
  25% in 0.0031 secs
  50% in 0.0038 secs
  75% in 0.0050 secs
  90% in 0.0067 secs
  95% in 0.0085 secs
  99% in 0.0326 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0019 secs, 0.3048 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0025 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0037 secs
  resp wait:	0.0050 secs, 0.0019 secs, 0.2988 secs
  resp read:	0.0001 secs, 0.0000 secs, 0.0126 secs

Status code distribution:
  [200]	19598 responses
```

```
hey -c 10 -z 10s http://localhost:2300/json

Summary:
  Total:	10.0013 secs
  Slowest:	0.1258 secs
  Fastest:	0.0003 secs
  Average:	0.0011 secs
  Requests/sec:	9060.0549

  Total data:	92226095 bytes
  Size/request:	1017 bytes

Response time histogram:
  0.000 [1]	|
  0.013 [90173]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.025 [184]	|
  0.038 [90]	|
  0.051 [52]	|
  0.063 [32]	|
  0.076 [26]	|
  0.088 [16]	|
  0.101 [10]	|
  0.113 [9]	|
  0.126 [19]	|


Latency distribution:
  10% in 0.0004 secs
  25% in 0.0005 secs
  50% in 0.0007 secs
  75% in 0.0011 secs
  90% in 0.0015 secs
  95% in 0.0019 secs
  99% in 0.0059 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0003 secs, 0.1258 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0133 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0030 secs
  resp wait:	0.0011 secs, 0.0003 secs, 0.1258 secs
  resp read:	0.0000 secs, 0.0000 secs, 0.0042 secs

Status code distribution:
  [200]	90612 responses
```

```
hey -c 10 -z 10s http://localhost:2300/string

Summary:
  Total:	10.0014 secs
  Slowest:	0.3345 secs
  Fastest:	0.0003 secs
  Average:	0.0011 secs
  Requests/sec:	9015.1305

  Total data:	178686010 bytes
  Size/request:	1981 bytes

Response time histogram:
  0.000 [1]	|
  0.034 [89985]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.067 [108]	|
  0.101 [44]	|
  0.134 [25]	|
  0.167 [0]	|
  0.201 [0]	|
  0.234 [0]	|
  0.268 [0]	|
  0.301 [0]	|
  0.335 [1]	|


Latency distribution:
  10% in 0.0004 secs
  25% in 0.0005 secs
  50% in 0.0007 secs
  75% in 0.0012 secs
  90% in 0.0015 secs
  95% in 0.0019 secs
  99% in 0.0054 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0000 secs, 0.0003 secs, 0.3345 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0026 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0029 secs
  resp wait:	0.0011 secs, 0.0003 secs, 0.3290 secs
  resp read:	0.0000 secs, 0.0000 secs, 0.0074 secs

Status code distribution:
  [200]	90164 responses
```

**Reply by konung** [reply]
That's great to hear. 

My MacBook is older - last generation Intel from 2019 with i9, hence slower results, probably.  

Having said that, even on default puma setup with 1 WEB_WORKER and default number of max threads (5 I think), it still 7x times faster.

**Comment by joshfester** [comment]
I just discovered Phlex recently, and it really excites me. I recently started using ViewComponent, but always felt like it wasn't quite what I wanted. Phlex seems to answer everything perfectly. Are there any benchmarks comparing it to ViewComponent and ERB + Helpers? That's what I'd be most interested in for a benchmark

**Comment by konung** [comment]
I agree Phlex could replace both Cells+Haml/Erb/Slim or ViewComponent + /Erb/Slim/Haml combinations. 
I'm waiting for Hanami 2.1 to be released in 2023 Q1, which will have Hanami::View to try these benchmarks. 

As far as in the context of Rails +  ActionView / View Component + Template - my gut feeling says that Phlex will be way faster than those combinations. Even before it's compiles as @joeldrapper  mentions. Simply because there are less layers, less orchestrations, and no template parsing. 

I haven't tried running Phlex in Rails context yet, but may try once I revisit this for Hanami 

On a side note:

I've tried ViewComponent. But can't get fully behind it . To mee it's seems like a more complicated version of Cells. I think Cells just "clicks" better for me, or maybe because I've been using them for so long. And Phlex in my subjective view is closer to Cells, than ViewComponent the way  you use it.

**Comment by joshfester** [comment]
Is Hanami::View going to be using Phlex? I agree it seems logical that it would be much faster than ViewComponent + ERB. I might take a stab at making some benchmarks. I'm constantly starting new Rails projects so it would be super helpful for me to have some data before I make Phlex my go-to.

I'm also very curious about benchmarking a view that's 100% Phlex (like from the application layout down to each component) vs a standard Rails ERB layout/page with some Phlex components in it.

I don't have any experience with Cells but it does look nice. I chose ViewComponent because they talk explicitly about performance. I'm always looking for gems like this that can make big improvements with minimal effort on my end. 

I also love that Phlex makes single-file components so easy. I know it's possible with ViewComponent, but then I'd have to use Rails' tag helpers, which I don't think are very performant

**Reply by joeldrapper** [reply]
I'd love to see some benchmarking. Phlex was about 10x faster than ActionView and 5x faster than ViewComponent when rendering small partials last time I checked. But it gets slower with lots of elements and attributes as it's not compiled yet.

It's about as optimised as it can be without compilation, so that's really the next hurdle for performance.

**Reply by joshfester** [reply]
Ah that is good to know. I will keep an eye on your compiler repo. Appreciate all the work you're doing on this, it fills a big need in the Rails stack!
